File: 序列化传输/std.py
N = 1000010

class Trie:
    def __init__(self):
        self.tr = [[0, 0] for _ in range(N)]
        self.idx = 0
        self.cnt = [0] * N

    def insert(self, s, id):
        p = 0
        for u in s:
            ch = int(u)
            if self.tr[p][ch] == 0:
                self.idx += 1
                self.tr[p][ch] = self.idx
            p = self.tr[p][ch]
            self.cnt[p] += 1

    def query(self, s):
        p = 0
        length = 0
        for u in s:
            ch = int(u)
            if self.tr[p][ch] != 0:
                p = self.tr[p][ch]
                length += 1
            else:
                break
        return length

    def remove(self, s):
        p = 0
        for u in s:
            ch = int(u)
            tmp = self.tr[p][ch]
            self.cnt[self.tr[p][ch]] -= 1
            if self.cnt[self.tr[p][ch]] == 0:
                self.tr[p][ch] = 0
            p = tmp

def Hex2Proto(s):
    res = ""
    for i in range(2, len(s)):
        num = 0
        if s[i].isdigit():
            num = int(s[i])
        else:
            num = ord(s[i]) - ord('A') + 10
        tmp = ""
        for j in range(4):
            tmp += str(num & 1)
            num >>= 1
        res += tmp[::-1]
    return res

def Proto2Binary(s):
    res = ""
    for i in range(len(s) - 1, -1, -8):
        for j in range(i - 6, i + 1):
            res += s[j]
    res = res[::-1]
    while len(res) > 0 and res[-1] == '0':
        res = res[:-1]
    res = res[::-1]
    return res

def solve():
    n = int(input())

    vec = ["0"]
    idx = 0
    s = 0

    mp = {}

    for _ in range(n):
        x, y = input().split()
        x = int(x)

        if x == 1:
            mp[y] = mp.get(y, 0) + 1
            Proto_y = Hex2Proto(y)
            Binary_y = Proto2Binary(Proto_y)

            s = T.query(Binary_y)
            T.insert(Binary_y, idx + 1)
            vec.append(Binary_y)

            print(mp[y], s)
        else:
            if y in mp and mp[y] > 0:
                mp[y] -= 1
                Proto_y = Hex2Proto(y)
                Binary_y = Proto2Binary(Proto_y)
                T.remove(Binary_y)

T = Trie()
solve()

File: 序列化传输/std.cpp

#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <chrono>
#define debug(x) std::cout<<#x<<": "<<x<<std::endl
using namespace std;

const int N = 10000100;

struct Trie{
    int tr[N][2],idx{0};  //trie树
    int cnt[N]; //cnt数组用于标记当前节点的个数

    void insert(string& s,int id){
        int p=0;
        for(auto u:s){
            int ch = u-'0';
            if(!tr[p][ch])
                tr[p][ch] = ++idx;
            p=tr[p][ch];
            cnt[p] ++;
        }
    }

    int query(string &s){  //返回值为下标
        int p=0, len = 0;
        for(auto u:s){
            int ch = u-'0';
            if(tr[p][ch]){
                p = tr[p][ch];
                ++len;
            }else{
                break;
            }
        }
        return len;
    }

    void remove(string &s){  //删除一个串
        int p = 0;
        for(auto u:s){
            int ch = u-'0';
            int tmp = tr[p][ch];
            cnt[tr[p][ch]] --;
            if(cnt[tr[p][ch]] == 0){
                tr[p][ch] = 0;
            }
            p = tmp;
        }
    }
}T;

string Hex2Proto(string &s){  //s=0x...
    string res = "";
    for(int i=2;i<(int)s.size();++i){
        int num = 0;
        if(isdigit(s[i])){
            num = s[i]-'0';
        }else{
            num = s[i]-'A'+10;
        }
        string tmp = "";
        for(int j=0;j<4;++j){
            tmp += ('0' + (num&1));
            num >>= 1;
        }
        reverse(tmp.begin(), tmp.end());
        res += tmp;
    }
    return res;
}

string Proto2Binary(string &s){
    string res = "";
    for(int i=s.size()-1;i>=0;i-=8){
        for(int j=i-6;j<=i;++j){
            res += s[j];
        }
    }
    reverse(res.begin(), res.end());
    while(res.size()&&res.back()=='0') res.pop_back();
    reverse(res.begin(), res.end());
    return res;
}

void solve(){
    int n; cin >> n;

    vector<string> vec(1, "0");     //存放所有的数值二进制过后的数据

    int idx = 0, s = 0;             //string数组下标, 相似长度

    unordered_map<string, int> mp; //存放每个编码的出现次数

    while(n--){
        int x;
        string y;
        cin >> x >> y;

        if(x == 1){
            mp[y] += 1;
            string Proto_y = Hex2Proto(y);
            string Binary_y = Proto2Binary(Proto_y);

            s = T.query(Binary_y);
            T.insert(Binary_y, ++idx);
            vec.push_back(Binary_y);

            cout << mp[y] << " " << s << endl;
        }else{
            if(mp.count(y) && mp[y] > 0) {
                mp[y] --;
                string Proto_y = Hex2Proto(y);
                string Binary_y = Proto2Binary(Proto_y);
                T.remove(Binary_y);
            }
        }
    }
}

signed main(){
    // ios::sync_with_stdio(false);
    // cin.tie(0);
    // cout.tie(0);
    // int _=1;
    // cin >> _;
    // while(_--)
    // int i;
    // cin >> i;
    // string InputFile = "in" + to_string(i) + ".txt";
    // string OutputFile = "out" + to_string(i) + ".txt";
    // freopen(InputFile.c_str(), "r", stdin);
    // freopen(OutputFile.c_str(), "w", stdout);
    // auto t0 = std::chrono::steady_clock::now();
    solve();
    // auto t1 = std::chrono::steady_clock::now();
    // auto dt = std::chrono::duration_cast<std::chrono::milliseconds>(t1 - t0);
    // cerr << dt.count() << "ms\n";
    
    return 0;
}